File: leetcode/0025-reverse-nodes-in-k-group/solution.go
/**
 * Author: 1chooo<hugo970217@gmail.com>
 * Problem: https://leetcode.com/problems/reverse-nodes-in-k-group
 * Runtime: 0ms (100.00%)
 */

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reverseKGroup(head *ListNode, k int) *ListNode {
	reverse := func(start, end *ListNode) *ListNode {
		var prev *ListNode
		current := start
		for current != end {
			nextNode := current.Next
			current.Next = prev
			prev = current
			current = nextNode
		}
		return prev
	}

	count := 0
	current := head
	for current != nil && count < k {
		current = current.Next
		count++
	}

	if count < k {
		return head
	}

	reversedHead := reverse(head, current)
	if head != nil {
		head.Next = reverseKGroup(current, k)
	}

	return reversedHead
}

File: leetcode/0025-reverse-nodes-in-k-group/solution.py
"""
* Author: 1chooo<hugo970217@gmail.com>
* Problem: https://leetcode.com/problems/reverse-nodes-in-k-group
* Runtime: 0ms (100.00%)
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def reverse(start, end):
            prev = None
            current = start
            while current != end:
                next_node = current.next
                current.next = prev
                prev = current
                current = next_node
            return prev
        
        count = 0
        current = head
        while current and count < k:
            current = current.next
            count += 1
        
        if count < k:
            return head
        
        reversed_head = reverse(head, current)
        head.next = self.reverseKGroup(current, k)
        
        return reversed_head
